src/util/circular_buffer.rs

```diff
@@ -0,0 +1,28 @@
+use std::collections::VecDeque;
+///A CircularBuffer that drops the oldest element inserted when full.
+pub struct CircularBuffer<A>{
+    queue:VecDeque<A>,
+    size:usize,
+}
+impl<A> CircularBuffer<A>{
+    pub fn new<B>(size:usize)->CircularBuffer<B>{
+        CircularBuffer{
+            queue: VecDeque::with_capacity(size),
+            size:size,
+        }
+    }
+    //front to back
+    ///Add a value to the CircularBuffer
+    pub fn push(&mut self,elem:A){
+        if self.queue.len() < self.size {
```

Comment:
This logic can be simplified by moving the `push_back` out of the if/else construct

Reply:
I have no idea why I did such a stupid error. Thanks for pointing it out!